### PR TITLE
Chronos: Adding `NBEATSForecaster` API

### PR DESCRIPTION
**Pull request by liangs6212**

### CODE REVIEW

**Comment by TheaperDeng** [comment]
please split to 2 PRs for forecaster and automodel respectively

**Comment by liangs6212** [comment]
# TODO: Modify parameter description.

File: python/chronos/src/bigdl/chronos/forecaster/nbeats_forecaster.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import torch
+from bigdl.chronos.forecaster.base_forecaster import BasePytorchForecaster
+from bigdl.chronos.model.nbeats_pytorch import model_creator, loss_creator, optimizer_creator
+
+
+class NBEATSForecaster(BasePytorchForecaster):
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/nbeats_forecaster.py:22]: `NBeatsForecaster`, please change the name in other files as well

File: python/chronos/src/bigdl/chronos/forecaster/nbeats_forecaster.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+               fashion. If True, the internal model will use an Orca Estimator.
+               If False, the internal model will use a pytorch model. The value
+               defaults to False.
+        :param workers_per_node: int, the number of worker you want to use.
+               The value defaults to 1. The param is only effective when
+               distributed is set to True.
+        :param distributed_backend: str, select from "torch_distributed" or
+               "horovod". The value defaults to "torch_distributed".
+        """
+        # TODO Currently generic mode not support distributed.
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/nbeats_forecaster.py:93]: raise a issue once this PR is merged.

Comment (TheaperDeng) [reply python/chronos/src/bigdl/chronos/forecaster/nbeats_forecaster.py:93]: For "not support distributed." do you mean it does not support `distributed=True`(real distributed mode) or the multi-process training we automatically triggered when user has more than 8 cores?

File: docs/readthedocs/source/doc/PythonAPI/Chronos/forecasters.rst [hunk 3]
```diff
@@ -42,6 +42,19 @@ Temporal Convolutional Networks (TCN) is a neural network that use convolutional
     :show-inheritance:
 
 
+NBEATSForecaster
+----------------------------------------
+
+:strong:`Please refer to` `BasePytorchForecaster <https://bigdl.readthedocs.io/en/latest/doc/PythonAPI/Chronos/forecasters.html#basepytorchforecaster>`__ :strong:`for other methods other than initialization`.
+
+Neural basis expansion analysis for interpretable time series forecasting (N-BEATS) is a deep neural architecture based on backward and forward residual links and a very deep stack of fully-connected layers. Nbeats can solve univariate time series point forecasting problems, being interpretable, and fast to train.
```
Comment (TheaperDeng) [inline docs/readthedocs/source/doc/PythonAPI/Chronos/forecasters.rst:50]: is this passage copied from anywhere else? if so, please reference.

File: python/chronos/test/bigdl/chronos/forecaster/test_nbeats_forecaster.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        forecaster.fit((train_data[0], train_data[1]), epochs=2)
+        # inference
+        nbeats_pred=forecaster.predict(test_data[0])
+        assert nbeats_pred.shape == test_data[1].shape
+        eva = forecaster.evaluate(test_data)
+        assert eva[0].shape == test_data[1].shape[1:]
+
+    def test_nbeats_forecaster_data_loader(self):
+        train_loader, _, _ = create_data(loader=True)
+        init_orca_context(cores=4, memory="4g")
```
Comment (TheaperDeng) [inline python/chronos/test/bigdl/chronos/forecaster/test_nbeats_forecaster.py:85]: I don't think we need `init_orca_context` and `stop_orca_context` here

File: python/chronos/test/bigdl/chronos/forecaster/test_nbeats_forecaster.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+        with pytest.raises(RuntimeError):
+            with tempfile.TemporaryDirectory() as tmp_dir_name:
+                ckpt_name = os.path.join(tmp_dir_name, "ckpt")
+                forecaster.save(ckpt_name)
+        with pytest.raises(RuntimeError):
+            forecaster.predict(test_data[0])
+        with pytest.raises(RuntimeError):
+            forecaster.evaluate(test_data)
+    
+    def test_nbeats_forecaster_xshard_input(self):
```
Comment (TheaperDeng) [inline python/chronos/test/bigdl/chronos/forecaster/test_nbeats_forecaster.py:149]: It seems that this ut is not for xshards?

File: python/chronos/test/bigdl/chronos/forecaster/test_nbeats_forecaster.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        stop_orca_context()
+
+    def test_nbeats_forecaster_dataloader_distributed(self):
+        train_data, _, _ = create_data(loader=True)
+        init_orca_context(cores=4, memory="4g")
+        forecaster = NBEATSForecaster(past_seq_len=24,
+                                      future_seq_len=5,
+                                      loss="mae",
+                                      lr=0.01,
+                                      distributed=False)
```
Comment (TheaperDeng) [inline python/chronos/test/bigdl/chronos/forecaster/test_nbeats_forecaster.py:220]: we should set this to True?

File: docs/readthedocs/source/doc/PythonAPI/Chronos/forecasters.rst [hunk 7]
```diff
@@ -42,6 +42,19 @@ Temporal Convolutional Networks (TCN) is a neural network that use convolutional
...
+
+.. automodule:: bigdl.chronos.forecaster.nbeats_forecaster
+    :members:
+    :undoc-membears:
+    :show-inheritance:
```
Comment (TheaperDeng) [inline docs/readthedocs/source/doc/PythonAPI/Chronos/forecasters.rst:55]: it seems that this API doc does not work: https://smalldl.readthedocs.io/en/nbeats_forecaster/doc/PythonAPI/Chronos/forecasters.html#nbeatsforecaster

File: python/chronos/src/bigdl/chronos/forecaster/nbeats_forecaster.py (hunk 2 above)
Comment (liangs6212) [reply python/chronos/src/bigdl/chronos/forecaster/nbeats_forecaster.py:93]: `distributed`=True is not supported, only happens when setting `stack_types`=("generic", "generic").
If `stack_types`=("generic", "trend" or "seasonality") the above problem will not occur.
I will modify the description to "not support distributed=True" in the next commit.

**Comment by liangs6212** [comment]
doc: https://smalldl.readthedocs.io/en/nbeats_forecaster/doc/PythonAPI/Chronos/forecasters.html#nbeatsforecaster

File: python/chronos/src/bigdl/chronos/forecaster/nbeats_forecaster.py (hunk 2 above)
Comment (TheaperDeng) [reply python/chronos/src/bigdl/chronos/forecaster/nbeats_forecaster.py]: It seems to be an orca issue (related to `find_unused_parameter` or something like it), we can raise an orca issue after this PR.

Comment (liangs6212) [reply python/chronos/src/bigdl/chronos/forecaster/nbeats_forecaster.py]: ok, l will describe the problem in detail.

**Review by TheaperDeng — APPROVED** [review]
LGTM

**Comment by liangs6212** [comment]
http://10.112.231.51:18888/view/BigDL-2.0-PR/job/BigDL-Chronos-PR-Validation/285/

**Pull request merged**